### PR TITLE
Support collaboration feature

### PR DESCRIPTION
**Pull request by msujew**
#### What it does

Related to https://github.com/eclipse-theia/theia/issues/2842 (maybe even closes it?)

Makes use of the newly designed `open-collaboration-protocol` (source code [here](https://github.com/TypeFox/open-collaboration-server)) to enable collaboration across users on the same workspace. The main ideas have already been outlined in https://github.com/eclipse-theia/theia/pull/12677.

This PR implements basic workspace and editor sharing. All guests will work on the host workspace and all edits to a file will be shared across all users. Presence information (i.e. cursor selection of other peers in the collaboration room) is also available.

#### How to test

1. Open Theia and click on the `share` status bar item. Open a new session (you may need to log in - this is only to give you a name). This will put a room code into your clipboard.
2. Open another Theia instance click on the status bar item again. Connect to a session and input the code from your clipboard.
3. Play around with the IDE. Please be aware that not everything is working 100% correctly and not everything might be addressed in this PR.

#### Follow-ups

There are a lot of improvements on this:
- Add more editors (like notebook editors)
- Improve the general UI/UX of this feature
- Add support for permission changes (especially read-only)
- Add support for collaborative debugging
- Add support for collaborative terminals
- Fix code https://github.com/eclipse-theia/theia/issues/14080

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/collaboration/src/browser/collaboration-frontend-contribution.ts
Comment (planger) [inline packages/collaboration/src/browser/collaboration-frontend-contribution.ts:134]:
```suggestion
                    label: nls.localize('theia/collaboration/makeWritable', 'Enable Workspace Editing'),
```

File: packages/collaboration/src/browser/collaboration-frontend-contribution.ts
Comment (planger) [inline packages/collaboration/src/browser/collaboration-frontend-contribution.ts:145]:
```suggestion
                    label: nls.localize('theia/collaboration/makeReadonly', 'Disable Workspace Editing'),
```

File: packages/collaboration/src/browser/collaboration-frontend-contribution.ts
Comment (planger) [inline packages/collaboration/src/browser/collaboration-frontend-contribution.ts:157]:
```suggestion
            label: nls.localize('theia/collaboration/stop', 'End Collaboration Session'),
```

File: packages/collaboration/src/browser/collaboration-frontend-contribution.ts
Comment (planger) [inline packages/collaboration/src/browser/collaboration-frontend-contribution.ts:158]:
```suggestion
            detail: nls.localize('theia/collaboration/stopDetail', 'Terminate the session, cease content sharing, and revoke access for others.'),
```

File: packages/collaboration/src/browser/collaboration-frontend-contribution.ts
Comment (planger) [inline packages/collaboration/src/browser/collaboration-frontend-contribution.ts:135]:
```suggestion
                    detail: nls.localize('theia/collaboration/makeWritableDetail', 'Allow collaborators to modify content in your workspace.'),
```

File: packages/collaboration/src/browser/collaboration-frontend-contribution.ts
Comment (planger) [inline packages/collaboration/src/browser/collaboration-frontend-contribution.ts:146]:
```suggestion
                    detail: nls.localize('theia/collaboration/makeReadonlyDetail', 'Restrict others from making changes to your workspace.'),
```

File: packages/collaboration/src/browser/collaboration-frontend-contribution.ts
Comment (planger) [inline packages/collaboration/src/browser/collaboration-frontend-contribution.ts:170]:
```suggestion
            detail: nls.localize('theia/collaboration/inviteDetail', 'Copy the invitation code for sharing it with others to join the session.'),
```

File: packages/collaboration/src/browser/collaboration-frontend-contribution.ts
Comment (planger) [inline packages/collaboration/src/browser/collaboration-frontend-contribution.ts:175]:
```suggestion
            label: nls.localize('theia/collaboration/stop', 'Leave Collaboration Session'),
```

File: packages/collaboration/src/browser/collaboration-frontend-contribution.ts
Comment (planger) [inline packages/collaboration/src/browser/collaboration-frontend-contribution.ts:286]:
```suggestion
            const makeReadonly = nls.localize('theia/collaboration/makeReadonly', 'Disable Workspace Editing');
```

File: packages/collaboration/src/browser/collaboration-instance.ts
Comment (planger) [inline packages/collaboration/src/browser/collaboration-instance.ts:176]:
```suggestion
                nls.localize('theia/collaboration/userWantsToJoin', "User '{0}' wants to join the collaboration room", user.email ? `${user.name} (${user.email})` : user.name)),
```

File: packages/collaboration/src/browser/collaboration-frontend-contribution.ts
Comment (planger) [inline packages/collaboration/src/browser/collaboration-frontend-contribution.ts:115]:
```suggestion
            label: nls.localize('theia/collaboration/joinRoom', 'Join Collaboration Session'),
```

File: packages/collaboration/src/browser/collaboration-frontend-contribution.ts
Comment (planger) [inline packages/collaboration/src/browser/collaboration-frontend-contribution.ts:109]:
```suggestion
                label: nls.localize('theia/collaboration/createRoom', 'Create New Collaboration Session'),
```

File: packages/collaboration/src/browser/collaboration-frontend-contribution.ts
Comment (planger) [inline packages/collaboration/src/browser/collaboration-frontend-contribution.ts:127]:
```suggestion
            detail: nls.localize('theia/collaboration/inviteDetail', 'Copy the invitation code for sharing it with others to join the session.'),
```

**Review by planger — COMMENTED** [review]
Great work on this feature, @msujew, and thanks for your contribution!

During testing, I encountered a few issues beyond those mentioned in the PR description, such as the broken save functionality. Additionally, there seems to be a problem with editor synchronization, as demonstrated in this video: [collabsession.webm](https://github.com/eclipse-theia/theia/assets/588090/fb31d680-1155-40c5-bf04-c582c89d2731). I encountered these issues while adding text alternately in different parts of the editor. The setup was using the latest `master` branch of the [open-collaboration-server](https://github.com/TypeFox/open-collaboration-server). Let me know if I can help with providing more specific steps to replicate these problems.

I've also added some minor inline comments regarding the wording and capitalization consistency of UI labels.

Regarding extending support to other types of editors (like notebooks or custom editors), this might be more relevant to the [open-collaboration-server repository](https://github.com/TypeFox/open-collaboration-server). However, I'm curious about the approach for integrating additional editors. It seems there are currently no generic message types in the [open-collaboration-protocol](https://github.com/TypeFox/open-collaboration-server/tree/main/packages/open-collaboration-protocol) that one could extend for syncing non-text editors. For Theia, I assume, we would then need to expose collaboration and connection events for other components to respond to new sessions or connection events, enabling them to communicate through these connections.

Looking ahead, should we consider extending the VS Code extension API or creating a VS Code extension that adapts the open-collaboration-server protocol to the [VSLS API](https://www.npmjs.com/package/vsls)? What are your thoughts on this?

Thanks again!

**Comment by msujew** [comment]
Hey @planger thanks for the feedback! It'll probably take a bit until I get to updating the PR, but just to answer your questions:

> Regarding extending support to other types of editors (like notebooks or custom editors), this might be more relevant to the [open-collaboration-server repository](https://github.com/TypeFox/open-collaboration-server). However, I'm curious about the approach for integrating additional editors. It seems there are currently no generic message types in the [open-collaboration-protocol](https://github.com/TypeFox/open-collaboration-server/tree/main/packages/open-collaboration-protocol) that one could extend for syncing non-text editors. For Theia, I assume, we would then need to expose collaboration and connection events for other components to respond to new sessions or connection events, enabling them to communicate through these connections.

There is a generic `Message` type (see [here](https://github.com/TypeFox/open-collaboration-server/blob/main/packages/open-collaboration-rpc/src/messages.ts)) that can be instantiated to support any kind of message for any kind of editor. The current editor messages are just using the class-based signatures for that, see [here](https://github.com/TypeFox/open-collaboration-server/blob/5f83597d850e7a6fed0dede82509c703fc01c637/packages/open-collaboration-rpc/src/messages.ts#L206-L225). Adding a new editor should be as simple as sending and responding to those message types:

```ts
connection.onRequest('graphicalEditor', (..args) => { ... });
...
connection.sendRequest('graphicalEditor', ...args);
```

No need to create a separate connection/session.

> Looking ahead, should we consider extending the VS Code extension API or creating a VS Code extension that adapts the open-collaboration-server protocol to the [VSLS API](https://www.npmjs.com/package/vsls)? What are your thoughts on this?

Yes, definitely. Though it likely needs to happen by effectively mocking the `vscode.extensions.getExtension('ms-vsliveshare.vsliveshare')` call, since replacing the import to `vsls` won't work due to bundling in extensions. It's definitely on my mind!

> Let me know if I can help with providing more specific steps to replicate these problems.

That would actually be great. I was aware that this issue could occur, but I wasn't able to replicate it.

**Comment by planger** [comment]
@msujew Thank you for your feedback and clarifying the question about the message!

>>  For Theia, I assume, we would then need to expose collaboration and connection events for other components to respond to new sessions or connection events, enabling them to communicate through these connections.

> ```connection.onRequest('graphicalEditor', (..args) => { ... });```
> No need to create a separate connection/session.

Great! I was wondering how to get access to the `connection` and react to session events from a Theia extension. Maybe I missed it, but there doesn't seem to be a way to register session listeners or obtain a `connection`, or is there?

>> Let me know if I can help with providing more specific steps to replicate these problems.

> That would actually be great. I was aware that this issue could occur, but I wasn't able to replicate it.

Sure, I'll try to reproduce it in the next days and get back to you.

Thanks again for pushing this excellent new feature forward!

**Comment by planger** [comment]
@msujew Hi again, I've now looking into reproducing this issue and for me it really seems very easy to reproduce. I've recorded a video for reproducing this with a fresh server session: 
[collabsession2.webm](https://github.com/eclipse-theia/theia/assets/588090/27cd9e98-bfec-45d4-9da1-87ddd39617bb)

I don't see anything suspicious in the logs of the collaboration server or Theia. If you want me to enable or add some logging, please let me know. I'm happy to help out in any way!

**Comment by msujew** [comment]
Hey @planger thanks for the help, I've integrated Yjs properly into the protocol now to support real editor syncing. Desyncs should no longer be possible now. The PR is now ready for a real review as most of the issues I had with it have been adressed.

**Comment by jonah-iden** [comment]
@msujew One smaller issue i noticed is "cursors" from other users in the session are not removed from a file when switching to/focusing another. The seem to only be removed when the other user is closing the file.

For example the host is working in file1 then switching to file2. The guest is now seeing the host cursor in both file1 and file2.

grafik

**Comment by jonah-iden** [comment]
Another thing: There is currently no feedback for a guest when joining a room and wating for the host to accept. Would be nice to have a progress message 'connecting to room' or 'waiting for host to let you join'

**Comment by jonah-iden** [comment]
The dirty state of files is also not shared currently right? But for the first version of this feature i think thats fine

**Comment by jonah-iden** [comment]
Files opened by the host before starting the session seem to not be synchronized at all

**Comment by msujew** [comment]
@jonah-iden Thanks for looking into this!

> One smaller issue i noticed is "cursors" from other users in the session are not removed from a file when switching to/focusing another. The seem to only be removed when the other user is closing the file.

Oh, yeah. I was wondering why my architecture for keeping track of all the peer cursors became so complicated. Having just a single cursor per peer makes stuff way easier :D

>Another thing: There is currently no feedback for a guest when joining a room and wating for the host to accept. Would be nice to have a progress message 'connecting to room' or 'waiting for host to let you join'

Good idea 👍 

> The dirty state of files is also not shared currently right? But for the first version of this feature i think thats fine

Right, I'm waiting for #13683 to be merged before tackling this.

> Files opened by the host before starting the session seem to not be synchronized at all. The user needs to reopen the file before it works

Cannot reproduce on my system:

https://github.com/eclipse-theia/theia/assets/4377073/e59298be-2def-4de3-9910-5881330e0e0a

File: packages/collaboration/src/browser/collaboration-frontend-contribution.ts
Comment (msujew) [reply packages/collaboration/src/browser/collaboration-frontend-contribution.ts]: FYI, I couldn't apply this change as `Disable Workspace Editing` is too long for notification buttons and always gets truncated.

**Comment by jonah-iden** [comment]
>Files opened by the host before starting the session seem to not be synchronized at all. The user needs to reopen the file before it works

seems to only happen on files that are automaticly opened on application startup and are visible when the guest connects. But im not totally sure.

Im also getting a weird effect sometimes when opening a file the host has already edited its just empty for the guest and only contains the content after the second time opening it

**Comment by planger** [comment]
@msujew Thank you very much for this excellent iteration! I can confirm that I can't reproduce the issue of the two editors getting out of sync anymore with plain consecutive editing on both sides :+1: 

I understand that save on the guest doesn't work yet (PR description). What seems to be a rather severe bug though is:
* Open file on host
* Open file on guest
* Host make some changes -> editor becomes dirty for both (great!)
* No guest closes the editor and selects "Don't save" -> file is reset to state on disk also for the host (pretty bad)
I'd expect the state to remain unchanged for the host.

Also, I observed some issues though with undo/redo. I managed to get into an inconsistent state between the two editors if I do a change on the left and a change on the right, then go back to left and undo, it seems to undo the change left *and* right (I feel it only should undo the change from the left). If now the right editor does an undo, I ended up in an inconsistent state. Can you reproduce this too?

**Comment by msujew** [comment]
> Thank you very much for this excellent iteration! I can confirm that I can't reproduce the issue of the two editors getting out of sync anymore with plain consecutive editing on both sides 👍

Glad to hear that! It took me quite a while to get that right 😅 

> Also, I observed some issues with undo/redo. I managed to get into an inconsistent state between the two editors if I do a change on the left and a change on the right, then go back to left and undo, it seems to undo the change left and right (I feel it only should undo the change from the left). If now the right editor does an undo, I ended up in an inconsistent state. Can you reproduce this too?

Yes, perfect reproduction steps, thank you :D I've tried replicating the behavior of VSCode Live Share, which actually undoes the latest changes (even if they're from someone else), but that doesn't work as well as I hoped it would. Anyway, I've adjusted it so that users can now only undo their own changes. I think that behavior is fine until I find a better solution for that.

> I understand that save on the guest doesn't work yet (PR description). What seems to be a rather severe bug though is:

Hm, I see. It took me a bit to figure out how to fix that but I think I found a well-working solution :)

I've pushed some changes that should improve on these problem areas. @jonah-iden I might have also fixed the first-load issue that you were experiencing as a drive-by, but can't guarantee anything.

**Comment by planger** [comment]
@msujew Thank you very much! a727293 indeed fixes the two raised issues:
1. Loss of all unsaved changes when the guest closes the dirty editor
2. The inconsistent state when one side is using undo

Excellent work!

I still feel that the undo/redo behavior -- even though it is consistent now -- is not ideal. I don't know how VS Code handles the undo/redo in editing sessions, but I personally find it very confusing if all participants share a common command stack and don't have separate command stacks (ie. undo/redoing only their changes), as it seems to be the case now.

Think of working on a larger file, making a change and then use undo. If someone else happens to have made a change between your last change and the undo (which easily happens if two or more persons work on a file), undo seemingly didn't do anything, even though it actually undid a foreign change out of sight (similarly it is weird for the other person whose change has been reverted and taken off their command stack). Even worse if you undo a few times... the first may undo your change, the second may undo another change that may have happened in the meantime.

Also, in the current implementation it also seems that the "undo granularity" is different for the one who made the change and for the others. The one who made the change has a word-by-word undo (as usual in Monaco), while all others seem to have a character by character undo (a bit cumbersome :-)).

However, feel free to leave any undo/redo improvements open for future PRs. Given the breadth of this change, the undo/redo behavior certainly may not be a blocker for having this merged, as long as it is consistent. :-)

**Comment by JonasHelming** [comment]
@msujew When do we plan to merge this?

**Comment by msujew** [comment]
@JonasHelming I've just updated this PR to the latest version of the collaboration packages. It no longer needs a local server but runs on the public instance that we're currently running on google cloud. We'll have a nice domain name for that service soon.

All that's needed to merge this is a final review and approval.

**Comment by tsmaeder** [comment]
@msujew the login doesn't really work for me in the electron version: the login page opens in an electron window which is unresponsive.

**Comment by msujew** [comment]
@tsmaeder I believe that due to https://github.com/eclipse-theia/theia/pull/13782, right? You can test this in the browser until it's merged.

**Comment by tsmaeder** [comment]
@msujew even in the browser version I'm the login page is not moving: in the log, I see a 400 error with the response text "Failed to perform simple login"

**Comment by msujew** [comment]
@tsmaeder This is ready for another review. Note that the current version is a `next` build and points towards the staging area of our server instance. The official update will follow in a few days.

**Comment by tsmaeder** [comment]
@msujew I can now log in, but on Electron/Windows, the login windows still open as extra Electron windows and can't be closed. Could you maybe rebase the PR so we can test that it works with the now merged PR about external links and Electron?

One UI suggestion: it would be neat if there was an indication which one of the users is the local user. If you have multiple users with the same name (thomas1, thomas2), it might be useful.

File: packages/collaboration/src/browser/collaboration-file-system-provider.ts
Comment (msujew) [reply packages/collaboration/src/browser/collaboration-file-system-provider.ts:80]: It's not great what we do here, but we don't get the string buffer directly from yjs, so we need to encode it to an array buffer to satisfy the API. I don't think this will lead to actual performance issues though.

File: packages/collaboration/src/browser/collaboration-file-system-provider.ts
Comment (msujew) [reply packages/collaboration/src/browser/collaboration-file-system-provider.ts:28]: Purely internal. Before leaving the system, the file path is transformed into a path that the collaboration protocol understands.

File: packages/collaboration/src/browser/collaboration-file-system-provider.ts
Comment (msujew) [reply packages/collaboration/src/browser/collaboration-file-system-provider.ts:111]: Yes, I tested this pretty extensively on both Ubuntu and Windows. It's also the same logic used in the [vscode extension for open collaboration](https://github.com/TypeFox/open-collaboration-tools/blob/16e7c68fcb8719c655f5f11dd0c82526e90288b3/packages/open-collaboration-vscode/src/collaboration-file-system.ts#L70-L75).

File: packages/collaboration/src/browser/collaboration-frontend-contribution.ts
Comment (msujew) [reply packages/collaboration/src/browser/collaboration-frontend-contribution.ts:99]: `opened` in this case doesn't mean that the explorer node is expanded. It just means that there is a workspace available (or not if the value is `false`). Since there is nothing to share if we don't have a workspace open, we don't show the option.

File: packages/collaboration/src/browser/collaboration-frontend-contribution.ts
Comment (msujew) [reply packages/collaboration/src/browser/collaboration-frontend-contribution.ts:180]: I agree - I've actually recently made the same change in the VS Code extension 👍

File: packages/collaboration/src/browser/collaboration-color-service.ts
Comment (msujew) [reply packages/collaboration/src/browser/collaboration-color-service.ts:75]: Added a comment about its source.

File: packages/collaboration/src/browser/collaboration-instance.ts
Comment (tsmaeder) [reply packages/collaboration/src/browser/collaboration-instance.ts:701]: But only because of `model.textEditorModel.getValue() !== newContent`. That we need this back stop tells me we don't really have a handle on the syncing logic. But fair enough...I henceforth reserve the right to summon you by your true name whenever the contents get out of sync ;-)

File: packages/collaboration/src/browser/collaboration-workspace-service.ts
Comment (msujew) [reply packages/collaboration/src/browser/collaboration-workspace-service.ts:26]: Done, see https://github.com/eclipse-theia/theia/issues/14080.

**Review by tsmaeder — APPROVED** [review]
I verified that the sign-in now opens a browser window. Let's merge this and continually improve as we go along.

File: packages/collaboration/src/browser/collaboration-frontend-contribution.ts
Comment (spoenemann) [inline packages/collaboration/src/browser/collaboration-frontend-contribution.ts:48]: We should change this to `https://api.open-collab.tools/` before merging.

**Comment by JonasHelming** [comment]
@msujew Congrats!

**Pull request merged**
